File: backend/app/schemas/post.py
"""互动领域相关模型。"""
from datetime import datetime
from typing import Optional

from pydantic import BaseModel, Field


class PostBase(BaseModel):
    """帖子基础字段，供复用定义。"""

    title: Optional[str] = Field(None, min_length=2, max_length=120, description="帖子标题")
    content_md: Optional[str] = Field(None, min_length=10, description="Markdown 内容")
    topic: Optional[str] = Field(None, max_length=120, description="所属话题")


class PostCreate(PostBase):
    """帖子创建请求体。"""

    title: str = Field(..., min_length=2, max_length=120, description="帖子标题")
    content_md: str = Field(..., min_length=10, description="Markdown 内容")
    submit_for_review: bool = Field(True, description="是否提交审核")


class PostUpdate(PostBase):
    """帖子更新请求体。"""

    status: Optional[str] = Field(None, description="状态")
    review_note: Optional[str] = Field(None, max_length=500, description="审核意见")


class PostRead(PostBase):
    """帖子响应结构。"""

    title: str = Field(..., min_length=2, max_length=120, description="帖子标题")
    content_md: str = Field(..., min_length=10, description="Markdown 内容")
    id: int
    status: str
    like_count: int
    favorite_count: int
    author_id: int
    reviewer_id: Optional[int]
    review_note: Optional[str]
    created_at: datetime
    updated_at: datetime

    class Config:
        orm_mode = True


class CommentCreate(BaseModel):
    """评论创建请求体。"""

    content: str = Field(..., min_length=2, max_length=500, description="评论内容")


class CommentRead(BaseModel):
    """评论响应体。"""

    id: int
    post_id: int
    author_id: int
    content: str
    status: str
    created_at: datetime

    class Config:
        orm_mode = True


class ReactionRequest(BaseModel):
    """点赞收藏请求体。"""

    reaction_type: str = Field(..., description="like 或 favorite")
    target_type: str = Field(..., description="post 或 resource")
    target_id: int = Field(..., ge=1, description="目标 ID")

File: frontend/src/api/post.ts
import { httpClient } from "./http";

export interface PostItem {
  id: number;
  title: string;
  content_md: string;
  topic: string | null;
  status: string;
  like_count: number;
  favorite_count: number;
  created_at: string;
  updated_at: string;
}

export interface CommentItem {
  id: number;
  post_id: number;
  author_id: number;
  content: string;
  status: string;
  created_at: string;
}

export interface PostQuery {
  status?: string;
  topic?: string;
  page?: number;
  pageSize?: number;
}

export interface PaginationMeta {
  page: number;
  page_size: number;
  total: number;
}

export interface PostListResult {
  items: PostItem[];
  meta: PaginationMeta;
}

export interface PostCreatePayload {
  title: string;
  content_md: string;
  topic?: string;
  submit_for_review?: boolean;
}

export interface CommentPayload {
  content: string;
}

export interface ReactionPayload {
  reaction_type: "like" | "favorite";
  target_type: "post" | "resource";
  target_id: number;
}

export async function fetchPosts(query: PostQuery = {}): Promise<PostListResult> {
  const { status, topic, page = 1, pageSize = 8 } = query;
  const { data } = await httpClient.get<{ data: PostItem[]; meta: PaginationMeta }>("/posts/", {
    params: {
      status,
      topic,
      page,
      page_size: pageSize
    }
  });
  return {
    items: data.data,
    meta: data.meta
  };
}

export async function createPost(payload: PostCreatePayload): Promise<PostItem> {
  const { data } = await httpClient.post<{ data: PostItem }>("/posts/", payload);
  return data.data;
}

export async function createComment(postId: number, payload: CommentPayload): Promise<CommentItem> {
  const { data } = await httpClient.post<{ data: CommentItem }>(`/posts/${postId}/comments`, payload);
  return data.data;
}

export async function fetchComments(postId: number): Promise<CommentItem[]> {
  const { data } = await httpClient.get<{ data: CommentItem[] }>(`/posts/${postId}/comments`);
  return data.data;
}

export async function sendReaction(payload: ReactionPayload): Promise<void> {
  await httpClient.post("/posts/reactions", payload);
}
